Mathlib/AlgebraicGeometry/Noetherian.lean

```diff
@@ -0,0 +1,335 @@
+    intro ⟨f, hf⟩
+    have : IsNoetherianRing Γ(X, X.basicOpen f) := hN ⟨f, hf⟩
+    have := U.prop.isLocalization_basicOpen f
+    have hEq := IsLocalization.algEquiv (.powers f) (Localization.Away f) Γ(X, X.basicOpen f)
+    exact isNoetherianRing_of_ringEquiv Γ(X, X.basicOpen f) hEq.symm.toRingEquiv
+  | hU => exact hS' _
+
+/-- A scheme is locally Noetherian if and only if it is covered by affine opens whose sections
+are noetherian rings.
+
+See [Har77], Proposition II.3.2. -/
+theorem isLocallyNoetherian_iff_of_iSup_eq_top {ι} {S : ι → X.affineOpens}
+    (hS : (⨆ i, S i : Opens X) = ⊤) :
+    IsLocallyNoetherian X ↔ ∀ i, IsNoetherianRing Γ(X, S i) :=
+  ⟨fun _ i => IsLocallyNoetherian.component_noetherian (S i),
+   isLocallyNoetherian_of_affine_cover hS⟩
+
+open CategoryTheory in
+/-- A version of `isLocallyNoetherian_iff_of_iSup_eq_top` using `Scheme.OpenCover`. -/
+theorem isLocallyNoetherian_iff_of_affine_openCover {𝒰 : Scheme.OpenCover.{v, u} X}
+    [∀ i, IsAffine (𝒰.obj i)] :
+    IsLocallyNoetherian X ↔
+    ∀ (i : 𝒰.J), IsNoetherianRing Γ(𝒰.obj i, ⊤) := by
+  constructor
+  · intro h i
+    let U := Scheme.Hom.opensRange (𝒰.map i)
+    have := h.component_noetherian ⟨U, isAffineOpen_opensRange _⟩
+    apply isNoetherianRing_of_ringEquiv (R := Γ(X, U))
+    apply CategoryTheory.Iso.commRingCatIsoToRingEquiv
+    exact (IsOpenImmersion.ΓIso_top (𝒰.map i)).symm
+  · intro hCNoeth
+    let fS i : X.affineOpens := ⟨Scheme.Hom.opensRange (𝒰.map i), isAffineOpen_opensRange _⟩
+    apply isLocallyNoetherian_of_affine_cover (S := fS)
+    rw [← Scheme.OpenCover.iSup_opensRange 𝒰]
+    intro i
+    apply isNoetherianRing_of_ringEquiv (R := Γ(𝒰.obj i, ⊤))
+    apply CategoryTheory.Iso.commRingCatIsoToRingEquiv
+    exact IsOpenImmersion.ΓIso_top (𝒰.map i)
+
+/-- If `𝒰` is an open cover of a scheme `X`, then `X` is locally noetherian if and only if
+`𝒰.obj i` are all locally noetherian. -/
+theorem isLocallyNoetherian_iff_openCover (𝒰 : Scheme.OpenCover X) :
+    IsLocallyNoetherian X ↔ ∀ (i : 𝒰.J), IsLocallyNoetherian (𝒰.obj i) := by
+  constructor
+  · intro h i
+    constructor
+    intro U
+    let V : X.affineOpens := ⟨𝒰.map i ''ᵁ U, IsAffineOpen.image_of_isOpenImmersion U.prop _⟩
+    suffices Γ(𝒰.obj i, U) ≅ Γ(X, V) by
+      convert isNoetherianRing_of_ringEquiv (R := Γ(X, V)) _
+      · apply CategoryTheory.Iso.commRingCatIsoToRingEquiv
+        exact this.symm
+      · exact h.component_noetherian V
+    rw [← Scheme.Hom.preimage_image_eq (𝒰.map i) U]
+    trans
+    · apply IsOpenImmersion.ΓIso
+    · suffices Scheme.Hom.opensRange (𝒰.map i) ⊓ V = V by
+        rw [this]
+      rw [← Opens.coe_inj]
+      rw [Opens.coe_inf, Scheme.Hom.opensRange_coe, IsOpenMap.functor_obj_coe,
+        Set.inter_eq_right, Set.image_subset_iff, Set.preimage_range]
+      exact Set.subset_univ _
+  · have : ∀ i, IsAffine (𝒰.affineRefinement.openCover.obj i) := by
+      intro i
+      rw [Scheme.AffineOpenCover.openCover_obj]
+      infer_instance
```

Comment:
Can you add
```
instance (𝒰 : X.AffineOpenCover) (i) : IsAffine (𝒰.openCover.obj i) :=
  inferInstanceAs (IsAffine (Spec (𝒰.obj i)))
```
to `Mathlib.AlgebraicGeometry.AffineScheme`? thanks. This will then be unnecessary.